cclunit-framework-source/doc/CCLUTMOCKING.md

```diff
@@ -0,0 +1,284 @@
+
+7. The table mocking APIs are not supported when called from within a reportwriter section.  It might be tempting to use a dummyt query to set up mock data from a record structure, but various mocking calls such as cclutCreateMockTable, cclutRemoveMockTable and cclutAddMockData cannot be executed within the context of a query (because the implementations execute queries). Use a for loop instead.
+ 
+8. Mocking the tdbexecute "reply_to" entity is unsupported under certain conditions, specifically if a call to free the "reply_to" entity is made just prior to calling tdbexecute.  If the scenario is truly necessary for a test, the best alternative is to define and use a subroutine for freeing the "reply_to" entity within the script and use a mock for that subroutine which does not actually perform the freeing of the "reply_to" entity.
+  
+## Example
+Below is an example of some of the APIs available in the CCL Unit Mocking framework along with some simple notes.
+
+Script-under-test:
+
+    drop program 1abc_mo_get_persons:dba go
+    create program 1abc_mo_get_persons:dba    
+    
+    declare newSize = i4 with protect, noconstant(0)
+    
+    select into "nl:"
+    from person p
+    plan p
+	order by p.person_id
+	detail
+		newSize = newSize + 1
+		stat = alterlist(reply->persons, newSize)
+		reply->persons[newSize].person_id = p.person_id
+		reply->persons[newSize].name_last = p.name_last
+		reply->persons[newSize].name_first = p.name_first
+		reply->persons[newSize].birth_dt_tm = p.birth_dt_tm
+	with nocounter
+    
+    end
+    go
+
+Test Code:
+
+    declare mock_table_person = vc with protect, noconstant("")
+    
+    ; Defining a mock person table.  The return value is the name of the mocked table.  
+    ; This can be useful to perform a select on the table after the script-under-test is complete
+    ; to verify (among other things) that an insert or a delete worked correctly.
+    set mock_table_person = cclutDefineMockTable("person", "person_id|name_last|name_first|birth_dt_tm",
+	    "f8|vc|vc|dq8")
+	
+	; Add a non-unique index to name_last
+	call cclutAddMockIndex("person", "name_last", FALSE)
+	
+	; Creates the mock table.  After this, it is available for DML statements.
+	call cclutCreateMockTable("person")
+	
+	; Create data for the table.
+	call cclutAddMockData("person", "1.0|Washington|George|01-JAN-1970 00:00") ;Will add George Washington 
+	call cclutAddMockData("person", "2.0|Adams|John|02-FEB-1971 11:11") ;Will add John Adams 
+	call cclutAddMockData("person", "3.0|Jefferson|\null|03-MAR-1972 22:22") ;Will add Jefferson (no first name) 
+	call cclutAddMockData("person", "4.0|Madison||04-APR-1973 10:33") ;Will add Madison (empty string for first name)
+	
```

Comment:
Given the lengthy comment about the return value from cclutDefineMockTable, one would expect this test to actually make use of the mock_table_person value, but it doesn't.

Although I advocate moving that discussion to the cclutDefineMockTable documentation, it seems reasonable to provide a demonstration somewhere.  It is somewhat contrived, but selecting the count of records from mock_table_person at this point would suffice.  It could even be set off with something like ";;; a contrived example to demonstrate a potential use for the return value from cclutDefineMockTable".

Reply:
Will do.